source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -0,0 +1,210 @@
+  const ir::Instruction* varInst = def_use_mgr_->GetDef(varId);
+  assert(varInst->opcode() == SpvOpVariable);
+  const uint32_t varTypeId = varInst->type_id();
+  const ir::Instruction* varTypeInst = def_use_mgr_->GetDef(varTypeId);
+  return varTypeInst->GetSingleWordInOperand(kSpvTypePointerStorageClass) ==
+      SpvStorageClassFunction;
+}
+
+void AggressiveDCEPass::AddStores(uint32_t ptrId) {
+  analysis::UseList* uses = def_use_mgr_->GetUses(ptrId);
+  if (uses == nullptr)
+    return;
+  for (auto u : *uses) {
+    SpvOp op = u.inst->opcode();
+    if (op == SpvOpStore)
+      worklist_.push(u.inst);
+    else if (op != SpvOpLoad)
+      AddStores(u.inst->result_id());
+  }
+}
+
+bool AggressiveDCEPass::AggressiveDCE(ir::Function* func) {
+  bool modified = false;
+  // Add non-local stores, block terminating and merge instructions
+  // to worklist. If function call encountered, return false, unmodified.
+  // TODO(): Improve in presence of function calls
+  for (auto& blk : *func) {
+    for (auto& inst : blk) {
+      switch (inst.opcode()) {
+      case SpvOpStore: {
+        uint32_t varId;
+        (void) GetPtr(&inst, &varId);
+        if (!IsLocalVar(varId)) {
+          worklist_.push(&inst);
+        }
+      } break;
+      case SpvOpLoopMerge:
+      case SpvOpSelectionMerge:
+      case SpvOpBranch:
+      case SpvOpBranchConditional:
+      case SpvOpSwitch: 
+      case SpvOpKill: 
+      case SpvOpUnreachable: 
+      case SpvOpReturn:
+      case SpvOpReturnValue: {
+        worklist_.push(&inst);
+      } break;
+      case SpvOpFunctionCall: {
+        return false;
+      } break;
+      default:
+        break;
+      }
+    }
+  }
+  // Perform closure on live instruction set. 
+  while (!worklist_.empty()) {
+    ir::Instruction* liveInst = worklist_.front();
+    live_insts_.insert(liveInst);
+    // Add all operand instructions if not already live
+    liveInst->ForEachInId([this](const uint32_t* iid) {
+      ir::Instruction* inInst = def_use_mgr_->GetDef(*iid);
+      if (live_insts_.find(inInst) == live_insts_.end())
+        worklist_.push(inInst);
+    });
+    // If local load, add all variable's stores if variable not already live
+    if (liveInst->opcode() == SpvOpLoad) {
+      uint32_t varId;
+      (void) GetPtr(liveInst, &varId);
+      if (IsLocalVar(varId)) {
+        if (live_local_vars_.find(varId) == live_local_vars_.end()) {
+          AddStores(varId);
+          live_local_vars_.insert(varId);
+        }
+      }
+    }
+    worklist_.pop();
+  }
+  // Mark all non-live instructions dead
+  for (auto& blk : *func) {
+    for (auto& inst : blk) {
+      if (live_insts_.find(&inst) != live_insts_.end())
+        continue;
+      dead_insts_.insert(&inst);
+    }
+  }
+  // Remove debug statements referencing dead instructions. This must
+  // be done before killing the instructions, otherwise there are dead
+  // objects in the def/use database.
+  for (auto& di : module_->debugs()) {
```

Comment:
I added logic to delete OpDecorate and OpDecorateId if their target id is dead. But I just added OpGroupDecorate to the live worklist (so that none of its ids will go dead) and added a TODO to deal with it nicely later.

Still need to add tests.